### ISSUE TITLE
Set up regression tests

### ISSUE THREAD

**Issue by atteggiani** [ACCESS-NRI/replace_landsurface#46]
It would be good to have some some regression tests with real data (alongside unit-testing) so we can have a way to check that the overall functionality is not changed when we refactor the codebase (it would also greatly help with the refactoring carried out within the unit-testing).

Since all the entry points for this packafe rely on data available on _Gadi_, the regression tests will have to be performed directly on _Gadi_.

I propose the following:
- Bash script to perform all the tests (to potentially run within a PBS job)
- Define bash script with `set -e` to exit as soon as an error is raised (so we are sure we fail the test if there are errors).
- Have 3-4 tests for each entry point in the package
- Data should be different for each test, with as much differentiation in the options/arguments as possible
- Try to use data already available on _Gadi_ and with a reasonably "fixed" location (so we avoid tests failing because the input data doesn't exist anymore)
- Try to use data that is as small as possible, or in general, the tests should complete in a reasonable amount of time (possibly within 10 minutes, better if we can reduce this further)
- To help with the above we might be able to execute all tests in parallel using different processes (this package's functions should use only 1 process at a time)

**Comment by aidanheerdegen** [comment]
A `bash` script is easiest, but it should be possible to utilise `pytest` in a later iteration, and that way have some control over what is run and good feedback about which tests fail.

**Comment by aidanheerdegen** [comment]
It would be good to target at least a subset of the regression tests to run on GitHub runners for fast CI testing. 

Options include:

1. Creating representative data on the fly, e.g. using Iris
2. Using a small subset of data, or data that is altered so that it compresses very well (small range of values) and convert to PP format
3. Upload data to S3 or ARDC object store and download (and cache) as part of the test suite

**Comment by atteggiani** [comment]
> A `bash` script is easiest, but it should be possible to utilise `pytest` in a later iteration, and that way have some control over what is run and good feedback about which tests fail.

Yes, we could have some test functions in a separate python test file that complete the whole processing and then compare the output data with the expected one.

**Comment by atteggiani** [comment]
The current entry points for the `replace_landsurface` package are:
- hres_eccb
- hres_ic

They both have the following input arguments:
- `--mask MASKFILE`
- `--file ORIGINAL_FILE`
- `--start START_TIME`
- `--type DRIVING_DATA_TYPE`

`hres_ic` also has:
- `--hres_ic HIGH_RESOLUTION_IC_FILE`

(The `hres_eccb` has the `--hres_ic` option in the script, but it doesn't really use it).

A few notes on the input arguments:
- It would be better if all the arguments in the multiple input arguments are all different (so we can maximize the independence of each test).
- I think we should try and cover all combinations for multiple-choice arguments (like `--type`).

For this reasons, I think I will need:

#### hres_eccb
2 different sets of valid inputs (all different), with the `--type` argument corresponding to `era5land` and `barra`, respectively (`astart` type doesn't really do anything in `hres_eccb`).

#### hres_ic
3 different sets of valid inputs (all different), with the `--type` argument corresponding to `era5land`, `barra` and `astart`, respectively.

@engelca can you please provide:
- the input sets based on the conditions specified above
- a file (filepath) corresponding to the expected output for each of these input sets

Thank you

**Comment by engelca** [comment]
@atteggiani 

The branch I put together to address the pylinting has an example of a single driving script.

Perhaps it would be a good place to start there
https://github.com/ACCESS-NRI/replace_landsurface/blob/3b1d48dd61d1da62e76c62022b33aca81b47ad6c/src/replace_landsurface_with_hres.py

I am busy with a few things at the moment but perhaps we can discuss this in the team meeting on Tuesday.

**Comment by atteggiani** [comment]
This issue is not related to `pylint` (**note:** There will be no linting/formatting carried out within this repo as specified in [this comment](https://github.com/ACCESS-NRI/replace_landsurface/issues/39#issuecomment-2655138949)).

I cannot see any of the input arguments mentioned above within the files in the branch you linked.
Would it be possible for you to provide with the set of input arguments in a comment of this issue?

I think it would be advisable to have discussions within this issue for multiple reasons. The main two that come to mind are:
- It can be more structured and the content related to Integration Tests is easily accessible all in one place (this issue)
- The content can be accessed also by people involved in the project but not present at the team meeting

Of course we can still have further discussions within the team meeting, but I would prefer the main discussion to happen here.

> [!IMPORTANT]
> Integration tests are very important to have a "baseline" of the code and make sure that any future changes don't alter the expected behaviour of the code. As a consequence, this issue is a blocker for many other issues related to python scripts changes.
